projects/packages/sync/src/class-settings.php

```diff
@@ -200,6 +201,84 @@ public static function update_settings( $new_settings ) {
 		$validated_settings = array_intersect_key( $new_settings, self::$valid_settings );
 		foreach ( $validated_settings as $setting => $value ) {
 
+			/**
+			 * Custom table migration logic.
+			 *
+			 * This needs to happen before the option is updated, to avoid race conditions where we update the option,
+			 * but haven't yet created the table or can't create it.
+			 *
+			 * On high-traffic sites this can lead to Sync trying to write in a non-existent table.
+			 *
+			 * So to avoid this, we're going to first try to initialize everything and then update the option.
+			 */
+			if ( 'custom_queue_table_enabled' === $setting ) {
```

Comment:
As discussed in Slack, it might be worth checking if an import is currently in progress and skip updating the setting in this case. I was also thinking if it would make sense to check if the site is "out of Sync" while migrating and think if want to add special handling for cases with very long queues

Reply:
Let's leave this for a separate PR, as we need to have a more proper solution for it that will over-complicate the current PR.